plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java

```diff
@@ -201,7 +213,39 @@ protected SchemaTableName createMaterializedViewStorageTable(ConnectorSession se
         String storageSchema = getStorageSchema(definition.getProperties()).orElse(viewName.getSchemaName());
         SchemaTableName storageTable = new SchemaTableName(storageSchema, storageTableName);
         List<ColumnMetadata> columns = definition.getColumns().stream()
-                .map(column -> new ColumnMetadata(column.getName(), typeManager.getType(column.getType())))
+                .map(column -> {
+                    Type type = typeManager.getType(column.getType());
+                    // Substitute types not supported by Iceberg.
+                    // Upon reading, the types will be coerced back to those requested by the user.
+                    if (type == TINYINT || type == SMALLINT) {
+                        type = INTEGER;
+                    }
+                    if (type instanceof CharType) {
+                        type = VARCHAR;
+                    }
+                    if (type instanceof TimeType timeType) {
+                        // Iceberg supports microsecond precision only
+                        type = timeType.getPrecision() <= 6
+                                ? TIME_MICROS
+                                : VARCHAR;
+                    }
+                    if (type instanceof TimeWithTimeZoneType) {
+                        type = VARCHAR;
+                    }
+                    if (type instanceof TimestampType timestampType) {
+                        // Iceberg supports microsecond precision only
+                        type = timestampType.getPrecision() <= 6
+                                ? TIMESTAMP_MICROS
+                                : VARCHAR;
+                    }
+                    if (type instanceof TimestampWithTimeZoneType timestampWithTimeZoneType) {
+                        // Iceberg does not store the time zone
+                        // TODO allow temporal partitioning on these columns, or MV property to
+                        //  drop zone info and use timestamptz directly
+                        type = VARCHAR;
+                    }
+                    return new ColumnMetadata(column.getName(), type);
```

Comment:
There are also custom types like ObjectIdType inside connectors (there was a requirement to support this scenario with the materialised views implementation developed at Starburst).
Would a coercion to VARCHAR and back work for any or most types ? Could we can make VARCHAR the final choice if nothing else matches ?

Reply:
> Would a coercion to VARCHAR and back work for any or most types ? 

can it be assumed safely?

for example, it wouldn't work for JSON

also, here in the connector we're accepting the data
engine didn't validate that our new types will be coercible back. accepting "some random types" may leave us writing data no-one can read

i am open to revisiting, but i am not ready to add a catch-all clause. 
in the worst case, a person defining MV will have to cast their ObjectIds to a varchar in the mv def